File: current_session/173.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator(object):

    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.arr = []

        def helper(node):
            if node.left:
                helper(node.left)
            self.arr.append(node.val)
            if node.right:
                helper(node.right)

        helper(root)
        # print(self.arr)
        self.cur = 0
        
    def next(self):
        """
        :rtype: int
        """
        self.cur += 1
        return self.arr[self.cur-1]

    def hasNext(self):
        """
        :rtype: bool
        """
        return self.cur < len(self.arr)



# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()

File: current_session/c++/173.cpp
#include <iostream>
#include <vector>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class BSTIterator {
public:
  BSTIterator(TreeNode* root) {
    dfs(root);
  }

  int next() {
    return valueArray[cur++];
  }

  bool hasNext() {
    return cur<valueArray.size();
  }
private:
  int cur = 0;
  vector<int> valueArray;
  void dfs(TreeNode* node) {
    if (node == nullptr) return;
    if (node->left != nullptr) dfs(node->left);
    valueArray.push_back(node->val);
    if (node->right != nullptr) dfs(node->right);
    return;
  }
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */